File: agents/orchestration/src/routing.mjs
import fs from 'node:fs';

function escapeRegex(value) {
  return value.replace(/[|\\{}()[\]^$+?.]/g, '\\$&');
}

export function globToRegex(pattern) {
  const normalized = pattern.replaceAll('\\', '/');
  let expression = '^';
  for (let index = 0; index < normalized.length; index += 1) {
    const character = normalized[index];
    if (character === '*' && normalized[index + 1] === '*') {
      index += 1;
      if (normalized[index + 1] === '/') {
        index += 1;
        expression += '(?:.*/)?';
      } else {
        expression += '.*';
      }
    } else if (character === '*') {
      expression += '[^/]*';
    } else if (character === '?') {
      expression += '[^/]';
    } else {
      expression += escapeRegex(character);
    }
  }
  return new RegExp(`${expression}$`, 'i');
}

function keywordMatches(text, keyword) {
  const escaped = escapeRegex(keyword.toLowerCase()).replaceAll('\\ ', '\\s+');
  return new RegExp(`(^|[^a-z0-9])${escaped}([^a-z0-9]|$)`, 'i').test(text);
}

export function matchRule(rule, taskText, changedFiles) {
  const normalizedTask = taskText.toLowerCase();
  const matchedKeywords = (rule.keywords ?? []).filter((keyword) => keywordMatches(normalizedTask, keyword));
  const matchedPaths = [];
  for (const pattern of rule.paths ?? []) {
    const matcher = globToRegex(pattern);
    for (const file of changedFiles) {
      if (matcher.test(file.replaceAll('\\', '/'))) matchedPaths.push({ pattern, file });
    }
  }
  return {
    matched: matchedKeywords.length > 0 || matchedPaths.length > 0,
    keywords: matchedKeywords,
    paths: matchedPaths
  };
}

export function loadRouting(filePath) {
  const config = JSON.parse(fs.readFileSync(filePath, 'utf8'));
  if (config.version !== 1 || !Array.isArray(config.routes) || !Array.isArray(config.risk_rules)) {
    throw new Error('routing.yaml must contain version 1 routes and risk_rules');
  }
  const ids = [...config.routes, ...config.risk_rules].map((rule) => rule.id);
  if (new Set(ids).size !== ids.length) throw new Error('Routing and risk rule IDs must be unique');
  return config;
}

export function loadCatalog(filePath) {
  const content = fs.readFileSync(filePath, 'utf8');
  const agents = [];
  for (const line of content.split(/\r?\n/)) {
    const match = line.match(/^  ([a-z0-9-]+):\s*$/);
    if (match) agents.push(match[1]);
  }
  if (!agents.length) throw new Error('No agents found in catalog.yaml');
  return agents;
}

export function matchRoutes(config, taskText, changedFiles) {
  return config.routes.flatMap((route) => {
    const reasons = matchRule(route, taskText, changedFiles);
    return reasons.matched ? [{ id: route.id, reasons, rule: route }] : [];
  });
}

File: agents/orchestration/src/routing.py
"""Routing configuration loading and deterministic rule matching."""

from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Any, Pattern


def glob_to_regex(pattern: str) -> Pattern[str]:
    """Translate the selector's small glob dialect to a compiled regex.

    Supports: * (any chars except /), ** (any path segment), ? (single char),
    [a-z] character classes, and {a,b} brace expansion (Recommendation #4).
    """
    normalized = pattern.replace("\\", "/")
    expression = "^"
    index = 0
    while index < len(normalized):
        character = normalized[index]
        if character == "{" :
            # Brace expansion: {a,b,c} -> (a|b|c) processed in-place.
            end_brace = normalized.find("}", index)
            if end_brace != -1:
                body = normalized[index + 1 : end_brace]
                parts = [re.escape(p) for p in body.split(",")]
                expression += "(" + "|".join(parts) + ")"
                index = end_brace + 1
                continue
        if character == "[" and index + 1 < len(normalized):
            # Character class: [abc] or [!a-z] negation.
            end_bracket = normalized.find("]", index)
            if end_bracket != -1:
                class_body = normalized[index + 1 : end_bracket]
                expression += f"[{class_body}]"
                index = end_bracket + 1
                continue
        if character == "*" and index + 1 < len(normalized) and normalized[index + 1] == "*":
            index += 1
            if index + 1 < len(normalized) and normalized[index + 1] == "/":
                index += 1
                expression += "(?:.*/)?"
            else:
                expression += ".*"
        elif character == "*":
            expression += "[^/]*"
        elif character == "?":
            expression += "[^/]"
        else:
            expression += re.escape(character)
        index += 1
    return re.compile(f"{expression}$", re.IGNORECASE)


def _keyword_matches(text: str, keyword: str) -> bool:
    escaped = re.escape(keyword.lower()).replace(r"\ ", r"\s+")
    return re.search(rf"(^|[^a-z0-9]){escaped}([^a-z0-9]|$)", text, re.IGNORECASE) is not None


def match_rule(rule: dict[str, Any], task_text: str, changed_files: list[str]) -> dict[str, Any]:
    normalized_task = task_text.lower()
    matched_keywords = [
        keyword for keyword in rule.get("keywords", []) if _keyword_matches(normalized_task, keyword)
    ]
    matched_paths: list[dict[str, str]] = []
    for pattern in rule.get("paths", []):
        matcher = glob_to_regex(pattern)
        for file_name in changed_files:
            normalized_file = file_name.replace("\\", "/")
            if matcher.search(normalized_file):
                matched_paths.append({"pattern": pattern, "file": file_name})
    return {
        "matched": bool(matched_keywords or matched_paths),
        "keywords": matched_keywords,
        "paths": matched_paths,
    }


def load_routing(file_path: Path) -> dict[str, Any]:
    with file_path.open("r", encoding="utf-8") as source:
        config = json.load(source)
    if (
        config.get("version") != 1
        or not isinstance(config.get("routes"), list)
        or not isinstance(config.get("risk_rules"), list)
    ):
        raise ValueError("routing.yaml must contain version 1 routes and risk_rules")
    ids = [rule.get("id") for rule in [*config["routes"], *config["risk_rules"]]]
    if len(set(ids)) != len(ids):
        raise ValueError("Routing and risk rule IDs must be unique")
    return config


def load_catalog(file_path: Path) -> list[str]:
    content = file_path.read_text(encoding="utf-8")
    agents = []
    for line in content.splitlines():
        match = re.match(r"^  ([a-z0-9-]+):\s*$", line)
        if match:
            agents.append(match.group(1))
    if not agents:
        raise ValueError("No agents found in catalog.yaml")
    return agents


def match_routes(
    config: dict[str, Any], task_text: str, changed_files: list[str]
) -> list[dict[str, Any]]:
    matches = []
    for route in config["routes"]:
        reasons = match_rule(route, task_text, changed_files)
        if reasons["matched"]:
            matches.append({"id": route["id"], "reasons": reasons, "rule": route})
    return matches



